### ISSUE TITLE
Wine support

### ISSUE THREAD

**Issue by Jookia** [nvaccess/nvda#11027]
### Is your feature request related to a problem? Please describe.

NVDA currently doesn't work on Wine meaning a lot of accessible Win32 apps can't be used accessibly in Wine on Linux.

### Describe the solution you'd like

I'd like to see NVDA running on Wine and reading out things on my screen.

### Describe alternatives you've considered

Hooking up Wine's accessibility implementation (or lack thereof) directly to Linux or Mac accessibility options would likely require the same amount of work as getting NVDA to work.

### Additional context

This might be a controversial bug given Linux support for NVDA has been ruled out, but I haven't seen much mention of Wine support in the issue tracker. Ideally NVDA wouldn't have to care about this but it would be nice if it would support Wine as a platform from a development perspective at least.

I recently spent time seeing what would need to be done to get Wine working in NVDA. It's a relatively big list but at the very minimum it comes down to implementing the IAccessible API and focus hooks. By stubbing a bunch of NVDA code you can even use a native oleacc.dll DLL to have NVDA read what's at the cursor in Firefox 37 on Wine.

**Comment by josephsl** [comment]
Hi, I think it would be best to ask Wine developers to tackle this, as the overall aim of Wine project is to provide a Windows API compatibility layer for Linux platforms. Doing so will open up many things not only for NVDA, but also for other assistive technologies that rely on MSAA support. Thanks.

**Comment by Jookia** [comment]
I did that this week too by opening a bunch of bugs focusing on the specific MSAA functionality needed and explaining more in the NVDA Wine bug. But some of the blocks for NVDA are unrelated to MSAA but for things that might not need to be implemented. For instance, NVDA is currently broken altogether because Wine doesn't support the CoInitialize function. Having NVDA notice that Wine is returning the error code E_NOTIMPL on some functions could allow it to silently ignore some errors and allow the program to run if those functions aren't absolutely necessary.

**Comment by josephsl** [comment]
Hi, so if you did modify NVDA and let it drop CoInitialize, does it even run then? Note that this is COM related, something Wine project should support better in order to let COM servers and clients interact better. Thanks.

**Comment by Jookia** [comment]
Yes, I can confirm that it works. I spent this week seeing how far I
could tweak NVDA for it to run in Wine and got to the point it could
cursor read Firefox web pages.

**Comment by josephsl** [comment]
Hi,

I see.

What I'm trying to get at is looking at this from long-term accessibility and advocacy standpoints. For years, Wine was praised for excellent compatibility with apps and games using Windows API features many apps use. The fact that we are talking about tweaking screen readers to work with Wine when the project doens't provide necessary modules speaks to the larger issue of advocating for accessibility of other platforms.

It can be argued that by modifying NVDA, it can bring immediate benefits such as improved market penetration by NVDA and ability to use Linux with help from Wine. At least this short-term solution may create an immediate marketing and public relations opportunity.

But it doesn't solve the fundamental problem: accessibility, or rather, lack of awareness of accessibility in various Linux projects. Awareness of accessibility among developers targeting Windows, mobile platforms, and the web is growing thanks to presence of screen readers, advocacy by screen reader user communities, and mandated standards. I imagine awareness about accessibility is growing amongst Linux projects (the other day there was an issue filed on wxPython project talking about accessibility of wxPython app on Linux). But in order for Linux accessibiilty to improve, it is far better to advocate for awareness of accessibility among more Linux projects, because Wine will not and cannot replace Windows architecture and culture unless Wine project community learns more about a need to do something about accessibility API's.

Speaking of architecture and culture, the best way to advocate for accessibility on Linux, in my opinion (for now) is pushing for native accessibility, or for that matter, making sure more Linux apps adopt accessibility API's present in Linux rather than let a foreign system do it with help from a compatibility layer. Not only this benefits Orca and friends which were designed to serve as native Linux screen reader to begin with (native as in using API tips and tricks provided by Linux kernel and subsystems), it also serves as a better marketing and public relations campaign - far better to advertise built-in support rather than through an intermediary that must go through another layer of indirection. To me, improving native Linux accessibility (especially Linux desktop GUI apps) is a more attractive proposal, as it can make Linux app accessibility future-proof for years.

For these reasons, I carefully vote to forward this issue to Wine project (that is, close this request for now, have a philosophical discussion with Wine project about awareness of accessibility, and have a talk with Linux app vendors about Linux accessibility API's).

Thanks.

**Comment by Jookia** [comment]
Having NVDA be more tolerant of Wine's bugs would cut down on what Wine
needs to do support NVDA functioning, even if in a limited form.

**Comment by josephsl** [comment]
Hi,

Still, I think that's a short-term solution. Although practical, it still doesn't address the bigger problem of awareness of accessibility by Linux app vendors. Sure, the proposed solution may allow NVDA to be used for various development purposes on Linux, but remember that we're talking about a foreign entity trying to get work done through a compatibility layer.

Another key thing to note: there are limits as to what screen readers, and in extension, the community around it can do about an app or an ecosystem. We the screen reader developers and community members can provide guidance when it comes to advising app vendors about accessibility (1password is a good example of that). IN worse case scenario, screen reader developers must write workarounds but if and only if the vendor of the app in question doesn't have time to improve accessibility or is unwilling to do so. In other words, regarding this case, I think it would be better to come back to the proposed solution if Wine developers are unwilling to learn more about benefits of implementing accessibility API support in Wine - after all, the best routes of persuasion are not code and performance, but hearts and attitudes.

Thanks.

**Comment by Jookia** [comment]
It's not meant to address the bigger issue of accessibility by Linux app
vendors. That's a separate issue to Wine accessibility.

This solution assumes Wine developers will implement accessibilty API
support in Wine, but NVDA requires other APIs too. Implementing the
accessibility API overnight will not make NVDA run on Wine.

**Comment by josephsl** [comment]
Hi,

We know that implementing accessibility API support overnight doesn't mean NVDA will run fully, and suppose Wine developers do implement necessary changes. But my overall question remains: are they even talking about adding compatibility layer for oleacc.dll and UIAutomationCore.dll, or for that matter, COM compatibility? If yes, we might consider this proposal, otherwise the best people to persuade would be Wine folks (I'm afraid).

Thanks.

**Comment by Jookia** [comment]
Does NVDA use UIA much? My impression was that only really MSAA support
needed to be added.

**Comment by josephsl** [comment]
Hi,

Yes, NVDA does use UIA, and UIA client/server support is needed if dealing with a Windows app with UIA server features running under Wine. As UI Automation is built on top of COM (Component Object Model), solid COM support is required to interact with UIA server apps. Lack of CoInitialize/CoInitializeEx in Wine indicates no solid COM support yet, making NVDA unusable as a UIA client, which is much better than MSAA to some respects.

Thanks.

**Comment by Jookia** [comment]
Oops, I was mistaking CoInitialize for CoEnableCallCancellation.
Yes Wine does support COM. I was thinking of taking a stab at
implementing parts of MSAA myself for Wine, at least the parts I'd be
able to do.

**Comment by josephsl** [comment]
Hi,

In this case, it would be best to talk to Wine developers about cooperating with AT-SPI developers to come up with some sort of a bridge between the two.

In order to support Windows screen readers (not only NVDA, but others), the following accessibility API's must be implemented:

* MSAA (Microsoft Active Accessibility)/IAccessible
* IAccessible2
* UI Automation (both server and client bits)

In addition, for full functionality, NVDA must be able to:

* Talk to MSAA and UIA server applications as a client
* Must be able to inject C/C++ DLL's inside certain apps for faster data retrieval (that's part of the infrastructure required to support Firefox and Chrome at least)

Thanks.

**Comment by josephsl** [comment]
Hi,

Speaking of additional accessibility API's:

* IAccessible2 is used to talk to Firefox, Chrome, and other apps implementing IAccessible2 server side components. IAccessible2 came out as enhanced version of MSAA and as an alternative to UI Automation.
* UI Automation is used in newer Windows releases in order to talk to newer apps. Although implementing IUIAutomation would suffice, for best results, at least IUIAutomation3 (Windows 8.1) is preferred for certain scenarios. NOte that with new feature updates, Windows 10 will bring changes to UI Automation.

Thanks.

**Comment by Jookia** [comment]
I don't believe an AT-SPI bridge would be as portable as supporting NVDA
given Wine works on ReactOS and macOS as well.

**Comment by josephsl** [comment]
Hi,

True. But at least looking into sing existing bridges might be helpful for a while before considering full native implementation.

Thanks.

**Comment by Jookia** [comment]
True. Thanks for all this feedback and input.

**Comment by tapper82** [comment]
Any news on this? I would love to test out ReactOS on a old laptop I have lying around.

**Comment by Jookia** [comment]
I think Wine recently added support for focus tracking with the mouse
pointer, but I'm not sure.

**Comment by josephsl** [comment]
Hi,

For ReactOS, I hope folks there added at least some MSAA bits in there, as NVDA (or for that matter, screen readers) cannot work without it (and hoops involved) because MSAA is a fallback if everything else doesn't work well. In theory you can run NVDA provided that ReactOS includes necessary bits (MSAA is just one of the components NVDA needs), but what I explained in 2020 still applies: accessibility requires collaboration and trust between users, advocates, screen reader vendors, app developers, and operating system vendors (I apologize if any of my posts from 2020 sounded offensive).

Thanks.

**Comment by webczat** [comment]
interestingly, trying out nvda on wine... i am able to install nvda. it can even, when running wine's notepad, announce the window title and that I am in an edit field, even though nothing else seems to work. isn't something happening on that front already? this is still very broken and very incomplete, for example when tabbing around the nvda installer it will not announce checkboxes or buttons but it will read their labels, so I literally can install it.

**Comment by Conmanx360** [comment]
There has been quite a bit of work put into upstreaming the UI Automation patches we have from Proton, so things should be improving for native UIA applications in upstream wine here soon.

The reason the edit control in notepad is able to be detected as an edit field is because we have implemented some of the methods for the default win32 MSAA proxy for that control. We're still missing MSAA proxy support for all the other win32 controls right now, but things that have native MSAA implementations should be working better. Wine's internet explorer as an example should be giving out accessibility information, since we now build wine-gecko with accessibility support as of version 2.47.3.

So, a lot of things are still missing, but things are improving and should continue to improve. Even if we're missing a specific feature, the amount of work needed to get it implemented by someone else should be a lot less once we've got everything upstream from Proton.

**Comment by alex19EP** [comment]
> There has been quite a bit of work put into upstreaming the UI Automation patches we have from Proton, so things should be improving for native UIA applications in upstream wine here soon.

that's very nice to hear, if i may ask, what's the proton's interest in UIA?

**Comment by Conmanx360** [comment]
> that's very nice to hear, if i may ask, what's the proton's interest in UIA?

The initial use case for UI Automation in Proton was detecting when edit controls gained keyboard focus so that we could pop up the on screen keyboard automatically on the Steam Deck. This is why edit is currently the only win32 MSAA proxy we have extra support for, and why `EVENT_OBJECT_FOCUS` is the only winevent we currently raise aside from `EVENT_OBJECT_CREATE`.

The second use case is an application called [Xalia](https://github.com/madewokherd/xalia) that is being developed, which will allow controller navigation of desktop applications using accessibility data. This will be useful for navigating game launchers using only the controller.

**Comment by savoptik** [comment]
Good day, colleagues.
Tell me, please, is there any new information on this topic?
We recently tried to run nvda in wine, but it was quite inglorious. There is a feeling that the focus does not move, and the same key is spoken every time you press it, as if there is a failure somewhere.

**Comment by 2-of-1** [comment]
Greetings.

Me, being blind my self, are also very interested in the topic. I use Windows, Mac and Linux for various tasks and really would like to see NVDA support in Wine. There are some handy tools in the windows world, which would be a great addition to my Mac workflow for example. So, Have my voice for the project. ^^

Best

Richard

**Comment by Adriani90** [comment]
@jeremyd2019, @bernhardu, @jhol, @gverm, @oliverklee sorry to tag you on this repository, but we have many blind and visually impaired users trying to work properly in Wine and it seems there are still some accessibility issues. I saw that you were active on the Wine repo which however does not seem to be officially managed. But I thought maybe you have some contact to the Wine team and we could establish a connection and see how we can bring forward the work on accessibility.
Any help from your side in this regard is very appreciated.

**Comment by oliverklee** [comment]
@Adriani90 No, I have no contact to the Wine time and I also am no Wine developer.

Instead of pinging people on this unofficial mirror, you might have a look on https://www.winehq.org/ to find the recommended way to contact the Wine developers about this.
